itests/__init__.py

```diff
@@ -0,0 +1,193 @@
+
+    """
+    ret = []
+    for key, value in conf.items():
+        ret.extend(_write_section(key, value, 1))
+    return '\n'.join(ret) + '\n'
+
+
+def _rm_if_empty(path):
+    """Convenience wrapper for removing empty directories."""
+    try:
+        path.rmdir()
+    except OSError:
+        return False
+    return True
+
+
+async def _poll_file(path, timeout=2, step=0.1, exists=True):
+    """Poll a file to wait for its creation or removal.
+
+    Arguments:
+        timeout (number):
+            Maximum time to wait in seconds.
+        step (number):
+            Polling interval in seconds.
+        exists (bool):
+            Set to True to check if a file exists, otherwise False.
+
+    Raises:
+        Exception:
+            If polling hits the timeout.
+
+    """
+    elapsed = 0
+    while path.exists() != exists:
+        await asyncio.sleep(step)
+        elapsed += step
+        if elapsed > timeout:
+            raise Exception(f'Timeout waiting for file creation: {path}')
+
+
+def _expanduser(path):
+    """Expand $HOME and ~ in paths.
+
+    This code may well become obsolete after job platforms work has been
+    merged.
+
+    """
+    path = str(path)
+    path = path.replace('$HOME', '~')
+    path = path.replace('${HOME}', '~')
+    path = Path(path).expanduser()
+    return path
+
+
+def _make_flow(run_dir, test_dir, conf, name=None):
+    """Construct a workflow on the filesystem."""
+    if not name:
+        name = str(uuid1())
+    flow_run_dir = (test_dir / name)
+    flow_run_dir.mkdir()
+    reg = str(flow_run_dir.relative_to(run_dir))
+    if isinstance(conf, dict):
+        conf = suiterc(conf)
+    with open((flow_run_dir / 'suite.rc'), 'w+') as suiterc_file:
+        suiterc_file.write(conf)
+    return reg
+
+
+def _make_scheduler(reg, is_restart=False, **opts):
+    """Return a scheduler object for a flow registration."""
+    # get options object
+    if is_restart:
+        options = RestartOptions(**opts)
+    else:
+        options = RunOptions(**opts)
+    # create workflow
+    return Scheduler(reg, options, is_restart=is_restart)
+
+
+def _flow(make_flow, make_scheduler, conf, name=None, is_restart=False,
+          **opts):
+    """Make a flow and return a scheduler object."""
+    reg = make_flow(conf, name=name)
+    return make_scheduler(reg, is_restart=is_restart, **opts)
+
+
+@asynccontextmanager
+async def _run_flow(run_dir, caplog, scheduler):
+    """Start a scheduler."""
+    success = True
+    contact = (run_dir / scheduler.suite / '.service' / 'contact')
+    try:
+        asyncio.get_event_loop().create_task(scheduler.start())
+        await _poll_file(contact)
+        yield caplog
+    except Exception as exc:
+        # something went wrong in the test
+        success = False
+        raise exc from None  # raise the exception so the test fails
+    finally:
+        await scheduler.shutdown(SchedulerStop(StopMode.AUTO.value))
```

Comment:
All schedulers start in a single process which makes tidying up afterwards a lot easier, just call the shutdown method.